File: Library/Library.Root/Control/LogBase.cs
using System;
using System.Collections.Specialized;

namespace Library.Root.Control
{
    /// <summary>
    /// Log page base class (net8.0 migration — System.Web.UI.Page removed)
    /// QueryString supplied via constructor parameter.
    /// </summary>
    public abstract class LogBase
    {
        protected Library.Root.Other.GenericCollection<Library.Root.Object.Log> _list;
        private readonly NameValueCollection _queryString;

        protected LogBase(NameValueCollection queryString)
        {
            _queryString = queryString;
            this.BindKey();
            this.BindData();
        }

        private void BindKey()
        {
            foreach (string _query in _queryString)
            {
                if (!string.IsNullOrEmpty(_query))
                {
                    switch (_query)
                    {
                        case "id":
                            this._key = Uri.UnescapeDataString(_queryString["id"]);
                            break;
                        case "key":
                            this._setupKey = _queryString["key"];
                            break;
                        case "page":
                            if (int.TryParse(_queryString["page"], out int pageNum))
                            {
                                this._Pageno = pageNum;
                            }
                            else
                            {
                                this._Pageno = 1;
                            }
                            break;
                    }
                }
            }
        }

        protected abstract void BindData();

        private string _key = string.Empty;
        public string Key
        {
            get { return _key; }
        }

        private int _Pageno = 0;
        public int PageNo
        {
            get { return _Pageno; }
            set { _Pageno = value; }
        }

        private string _setupKey = string.Empty;
        public string SetupKey
        {
            get { return _setupKey; }
        }

        public string KeyDesc
        {
            get
            {
                if (_list.TotalRow > 0)
                {
                    return _list.Data[0].KeyDesc;
                }
                else
                {
                    return string.Empty;
                }
            }
        }

        public abstract string LogTitle { get; }
        public abstract string LogPage { get; }

        public string NormalTitle
        {
            get { return this.LogTitle; }
        }

        public string DisplayTitle
        {
            get { return this.LogTitle + " Audit Trail"; }
        }

        public string GenerateList
        {
            get { return this.LogPage + "?id=" + Uri.EscapeDataString(this.Key) + "&key=" + this.SetupKey + "&page=" + this.PageNo; }
        }
    }
}

File: Library/Library.Root/Control/LogBase.vb
Imports System.Collections.Specialized

Namespace Control
    ''' <summary>
    ''' Log page base class (net8.0 migration — System.Web.UI.Page removed)
    ''' QueryString supplied via constructor parameter.
    ''' </summary>
    Public MustInherit Class LogBase

        Protected _list As Other.GenericCollection(Of [Object].Log)
        Private ReadOnly _queryString As NameValueCollection

        Protected Sub New(ByVal queryString As NameValueCollection)
            _queryString = queryString
            Me.BindKey()
            Me.BindData()
        End Sub

        Private Sub BindKey()
            For Each _query As String In _queryString
                If Not String.IsNullOrEmpty(_query) Then
                    Select Case _query
                        Case "id"
                            Me._key = Uri.UnescapeDataString(_queryString("id"))
                        Case "key"
                            Me._setupKey = _queryString("key")
                        Case "page"
                            If IsNumeric(_queryString("page")) Then
                                Me._Pageno = _queryString("page")
                            Else
                                Me._Pageno = 1
                            End If
                    End Select
                End If
            Next
        End Sub

        Protected MustOverride Sub BindData()

        Private _key As String = String.Empty
        Public ReadOnly Property Key() As String
            Get
                Return _key
            End Get
        End Property

        Private _Pageno As Integer = 0
        Public Property PageNo() As Integer
            Get
                Return _Pageno
            End Get
            Set(ByVal value As Integer)
                _Pageno = value
            End Set
        End Property

        Private _setupKey As String = String.Empty
        Public ReadOnly Property SetupKey() As String
            Get
                Return _setupKey
            End Get
        End Property

        Public ReadOnly Property KeyDesc() As String
            Get
                If _list.TotalRow > 0 Then
                    Return _list.Data.Item(0).KeyDesc
                Else
                    Return String.Empty
                End If
            End Get
        End Property

        Public MustOverride ReadOnly Property LogTitle() As String
        Public MustOverride ReadOnly Property LogPage() As String

        Public ReadOnly Property NormalTitle() As String
            Get
                Return Me.LogTitle
            End Get
        End Property

        Public ReadOnly Property DisplayTitle() As String
            Get
                Return Me.LogTitle & " Audit Trail"
            End Get
        End Property

        Public ReadOnly Property GenerateList() As String
            Get
                Return Me.LogPage & "?id=" & Uri.EscapeDataString(Me.Key) & "&key=" & Me.SetupKey & "&page=" & Me.PageNo
            End Get
        End Property

    End Class
End Namespace
